Incorrect handling of external function call with vector of double as argument

## Description

The bug was initially discovered while testing [SLEEF](https://github.com/shibatch/sleef) library and minimized to a simple external function call with vector of double as argument. I would expect that assert never going to happen even with undefined reference to the definition of function. But since it happened already, I tried to figure out what exactly happens.

The comment in code says as follows

> // normal external function handling path
> // allocate 128 bits for each argument (+return value) to support fp80's;

So, it should be enough to store 128-bit arguments in there. And when it comes time to store `ConstantExpr` into memory, `toMemory` doesn't handle the case of 128-bit expression and calls assert without meaningful info. Therefore, I guess appropriate API for transforming `APInt` into raw bytes is missing and could be implemented based on [llvm::StoreIntToMemory](https://github.com/llvm/llvm-project/blob/main/llvm/lib/Support/APInt.cpp#L3016). Would you consider these changes?

## Steps to reproduce

- Create file `main.c`
```c
#include <immintrin.h>

typedef __m128d vdouble;

int f(vdouble d);

int main() {
        vdouble d = {};
        f(d);
        return 0;
}
```
- Compile `main.c`
```
clang -emit-llvm -c -g -O0 -Xclang -disable-O0-optnone main.c
```
- Run KLEE
```
klee main.bc
```
- Ensure that assert fails
```
/mnt/c/Work/klee/cmake-build-debug/bin/klee main.bc
KLEE: output directory is "/home/pavely/wllvm/sleef/build/klee-out-26"
KLEE: Using Z3 solver backend
KLEE: WARNING: undefined reference to function: f
klee: /mnt/c/Work/klee/lib/Expr/Expr.cpp:356: void klee::ConstantExpr::toMemory(void *): Assertion `0 && "invalid type"' failed.
 #0 0x00007f72fb9d84ff llvm::sys::PrintStackTrace(llvm::raw_ostream&) (/lib/x86_64-linux-gnu/libLLVM-10.so.1+0x9814ff)
 #1 0x00007f72fb9d6782 llvm::sys::RunSignalHandlers() (/lib/x86_64-linux-gnu/libLLVM-10.so.1+0x97f782)
 #2 0x00007f72fb9d8ac5 (/lib/x86_64-linux-gnu/libLLVM-10.so.1+0x981ac5)
 #3 0x00007f72fab5a0c0 (/lib/x86_64-linux-gnu/libc.so.6+0x430c0)
 #4 0x00007f72fab5a03b raise /build/glibc-sMfBJT/glibc-2.31/signal/../sysdeps/unix/sysv/linux/raise.c:51:1
 #5 0x00007f72fab39859 abort /build/glibc-sMfBJT/glibc-2.31/stdlib/abort.c:81:7
 #6 0x00007f72fab39729 get_sysdep_segment_value /build/glibc-sMfBJT/glibc-2.31/intl/loadmsgcat.c:509:8
 #7 0x00007f72fab39729 _nl_load_domain /build/glibc-sMfBJT/glibc-2.31/intl/loadmsgcat.c:970:34
 #8 0x00007f72fab4b006 (/lib/x86_64-linux-gnu/libc.so.6+0x34006)
 #9 0x00000000005a2654 klee::ConstantExpr::toMemory(void*) /mnt/c/Work/klee/lib/Expr/Expr.cpp:0:12
#10 0x000000000044a288 klee::Executor::callExternalFunction(klee::ExecutionState&, klee::KInstruction*, llvm::Function*, std::vector<klee::ref<klee::Expr>, std::allocator<klee::ref<klee::Expr> > >&) /mnt/c/Work/klee/lib/Core/Executor.cpp:3830:23
#11 0x00000000004485fe klee::Executor::executeCall(klee::ExecutionState&, klee::KInstruction*, llvm::Function*, std::vector<klee::ref<klee::Expr>, std::allocator<klee::ref<klee::Expr> > >&) /mnt/c/Work/klee/lib/Core/Executor.cpp:1667:7
#12 0x000000000044db8c klee::Executor::executeInstruction(klee::ExecutionState&, klee::KInstruction*) /mnt/c/Work/klee/lib/Core/Executor.cpp:2444:5
#13 0x00000000004562b9 klee::Executor::run(klee::ExecutionState&) /mnt/c/Work/klee/lib/Core/Executor.cpp:3524:5
#14 0x00000000004597b4 klee::Executor::runFunctionAsMain(llvm::Function*, int, char**, char**) /mnt/c/Work/klee/lib/Core/Executor.cpp:4436:3
#15 0x00000000004263eb main /mnt/c/Work/klee/tools/klee/main.cpp:1532:19
#16 0x00007f72fab3b0b3 __libc_start_main /build/glibc-sMfBJT/glibc-2.31/csu/../csu/libc-start.c:342:3
#17 0x0000000000420dfe _start (/mnt/c/Work/klee/cmake-build-debug/bin/klee+0x420dfe)
```

## Environment
```
/mnt/c/Work/klee/cmake-build-debug/bin/klee -version
KLEE 2.3-pre (https://klee.github.io)
  Build mode: Debug (Asserts: ON)
  Build revision: 71c1c45fe266d73e307f422f09f2afdc41c70ec3

LLVM (http://llvm.org/):
  LLVM version 10.0.0

  Optimized build.
  Default target: x86_64-pc-linux-gnu
  Host CPU: znver1
```
```
clang --version
clang version 10.0.0-4ubuntu1
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
```
```
uname -srm
Linux 5.10.16.3-microsoft-standard-WSL2 x86_64
```

--- a/include/klee/Expr/Expr.h
+++ b/include/klee/Expr/Expr.h
@@ -95,7 +95,9 @@ public:
 
   /// The type of an expression is simply its width, in bits. 
   typedef unsigned Width; 
-  
+
+  // NOTE: The prefix "Int" in no way implies the integer type of expression.
+  // For example, Int64 can indicate i64, double or <2 * i32> in different cases.
   static const Width InvalidWidth = 0;
   static const Width Bool = 1;
   static const Width Int8 = 8;
@@ -103,6 +105,10 @@ public:
   static const Width Int32 = 32;
   static const Width Int64 = 64;
   static const Width Fl80 = 80;
+  static const Width Int128 = 128;
+  static const Width Int256 = 256;
+  static const Width Int512 = 512;
+  static const Width MaxWidth = Int512;
   
 
   enum Kind {

--- a/lib/Core/Executor.cpp
+++ b/lib/Core/Executor.cpp
@@ -3810,11 +3810,13 @@ void Executor::callExternalFunction(ExecutionState &state,
   }
 
   // normal external function handling path
-  // allocate 128 bits for each argument (+return value) to support fp80's;
+  // allocate 512 bits for each argument (+return value) to support
+  // fp80's and SIMD vectors as parameters for external calls;
   // we could iterate through all the arguments first and determine the exact
   // size we need, but this is faster, and the memory usage isn't significant.
-  uint64_t *args = (uint64_t*) alloca(2*sizeof(*args) * (arguments.size() + 1));
-  memset(args, 0, 2 * sizeof(*args) * (arguments.size() + 1));
+  size_t allocatedBytes = Expr::MaxWidth / 8 * (arguments.size() + 1);
+  uint64_t *args = (uint64_t*) alloca(allocatedBytes);
+  memset(args, 0, allocatedBytes);
   unsigned wordIndex = 2;
   for (std::vector<ref<Expr> >::iterator ai = arguments.begin(), 
        ae = arguments.end(); ai!=ae; ++ai) {

--- a/lib/Expr/Expr.cpp
+++ b/lib/Expr/Expr.cpp
@@ -16,6 +16,7 @@
 // Core. If we need to do arithmetic, we probably want to use APInt.
 #include "klee/Support/IntEvaluation.h"
 
+#include "llvm/ADT/ArrayRef.h"
 #include "llvm/ADT/Hashing.h"
 #if LLVM_VERSION_CODE >= LLVM_VERSION(13, 0)
 #include "llvm/ADT/StringExtras.h"
@@ -311,6 +312,9 @@ void Expr::printWidth(llvm::raw_ostream &os, Width width) {
   case Expr::Int32: os << "Expr::Int32"; break;
   case Expr::Int64: os << "Expr::Int64"; break;
   case Expr::Fl80: os << "Expr::Fl80"; break;
+  case Expr::Int128: os << "Expr::Int128"; break;
+  case Expr::Int256: os << "Expr::Int256"; break;
+  case Expr::Int512: os << "Expr::Int512"; break;
   default: os << "<invalid type: " << (unsigned) width << ">";
   }
 }
@@ -336,23 +340,32 @@ void Expr::dump() const {
 
 ref<Expr> ConstantExpr::fromMemory(void *address, Width width) {
   switch (width) {
+  default: assert(0 && "invalid width");
   case  Expr::Bool: return ConstantExpr::create(*(( uint8_t*) address), width);
   case  Expr::Int8: return ConstantExpr::create(*(( uint8_t*) address), width);
   case Expr::Int16: return ConstantExpr::create(*((uint16_t*) address), width);
   case Expr::Int32: return ConstantExpr::create(*((uint32_t*) address), width);
   case Expr::Int64: return ConstantExpr::create(*((uint64_t*) address), width);
   // FIXME: what about machines without x87 support?
-  default:
-    return ConstantExpr::alloc(
-        llvm::APInt(width,
-                    (width + llvm::APFloatBase::integerPartWidth - 1) /
-                        llvm::APFloatBase::integerPartWidth,
-                    (const uint64_t *)address));
+  case Expr::Fl80: {
+    size_t numWords = (width + llvm::APFloatBase::integerPartWidth - 1) /
+                    llvm::APFloatBase::integerPartWidth;
+    return ConstantExpr::alloc(llvm::APInt(
+        width, llvm::ArrayRef<uint64_t>((const uint64_t *)address, numWords)));
+  }
+  case Expr::Int128:
+  case Expr::Int256:
+  case Expr::Int512: {
+    size_t numWords = width / APInt::APINT_BITS_PER_WORD;
+    return ConstantExpr::alloc(llvm::APInt(
+        width, llvm::ArrayRef<uint64_t>((const uint64_t *)address, numWords)));
+  }
   }
 }
 
 void ConstantExpr::toMemory(void *address) {
-  switch (getWidth()) {
+  auto width = getWidth();
+  switch (width) {
   default: assert(0 && "invalid type");
   case  Expr::Bool: *(( uint8_t*) address) = getZExtValue(1); break;
   case  Expr::Int8: *(( uint8_t*) address) = getZExtValue(8); break;
@@ -363,6 +376,10 @@ void ConstantExpr::toMemory(void *address) {
   case Expr::Fl80:
     *((long double*) address) = *(const long double*) value.getRawData();
     break;
+  case Expr::Int128:
+  case Expr::Int256:
+  case Expr::Int512:
+      memcpy(address, value.getRawData(), width / 8);
   }
 }
 

--- a/test/VectorInstructions/external_call.c
+++ b/test/VectorInstructions/external_call.c
@@ -1,0 +1,79 @@
+// RUN: %clang -DDYNAMIC_LIBRARY=1 %s -shared -o %t1.so
+// RUN: %clang %s -emit-llvm %O0opt -g -c -o %t1.bc
+// RUN: rm -rf %t.klee-out
+// NOTE: Have to pass `--optimize=false` to avoid vector operations being
+// constant folded away.
+// RUN: export LD_PRELOAD=%t1.so
+// RUN: export DYLD_INSERT_LIBRARIES=%t1.so
+// RUN: %klee --output-dir=%t.klee-out --optimize=false --exit-on-error --external-calls=all %t1.bc
+
+#include <stdint.h>
+
+typedef uint32_t v4ui __attribute__((vector_size(16)));  // 128-bit vector
+typedef uint32_t v8ui __attribute__((vector_size(32)));  // 256-bit vector
+typedef uint32_t v16ui __attribute__((vector_size(64))); // 512-bit vector
+
+v4ui call4(v4ui v);
+
+v8ui call8(v8ui v);
+
+v16ui call16(v16ui v);
+
+int call_mixed(v16ui v16, v8ui v8, v4ui v4);
+
+#ifdef DYNAMIC_LIBRARY
+
+v4ui call4(v4ui v) {
+  return v;
+}
+
+v8ui call8(v8ui v) {
+  return v;
+}
+
+v16ui call16(v16ui v) {
+  return v;
+}
+
+int call_mixed(v16ui v16, v8ui v8, v4ui v4) {
+  return v16[15] + v8[7] + v4[3];
+}
+
+#else
+
+#include "assert.h"
+#include "klee/klee.h"
+
+int main() {
+  v4ui v4 = {1, 2, 3, 4};
+  {
+    v4ui r = call4(v4);
+    for (int i = 0; i < 4; ++i) {
+      klee_assert(r[i] == v4[i]);
+    }
+  }
+
+  v8ui v8 = {1, 2, 3, 4, 5, 6, 7, 8};
+  {
+    v8ui r = call8(v8);
+    for (int i = 0; i < 8; ++i) {
+      klee_assert(r[i] == v8[i]);
+    }
+  }
+
+  v16ui v16 = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16};
+  {
+    v16ui r = call16(v16);
+    for (int i = 0; i < 16; ++i) {
+      klee_assert(r[i] == v16[i]);
+    }
+  }
+
+  {
+    int r = call_mixed(v16, v8, v4);
+    klee_assert(r == 28);
+  }
+  return 0;
+}
+
+#endif

